src/kernel/panic.zig

```diff
@@ -1,13 +1,148 @@
+    // Exit if we haven't loaded all debug modules
+    if (mem_profile.boot_modules.len < 1)
+        return;
+    var kmap_start: u32 = 0;
+    var kmap_end: u32 = 0;
+    for (mem_profile.boot_modules) |module| {
+        const mod_start = mem.physToVirt(module.mod_start);
+        const mod_end = mem.physToVirt(module.mod_end);
+        const mod_str_ptr = mem.physToVirt(@intToPtr([*]u8, module.cmdline));
+        const mod_str = mod_str_ptr[0..strlen(mod_str_ptr)];
+        if (std.mem.eql(u8, mod_str, "kernel.map")) {
+            kmap_start = mod_start;
+            kmap_end = mod_end;
+            break;
+        }
+    }
+    std.debug.assert(kmap_start != 0 and kmap_end != 0);
+
+    var syms = SymbolMap.init(allocator);
+    var i: u32 = kmap_start;
+    while (i < kmap_end) {
+        while (std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        // Parse the address from the first column
+        const addr_start = i;
+        // Each column is separated by one or more whitespace characters, so read until the first one
+        while (!std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        // Convert to an integer
+        const addr_str = @intToPtr([*]u8, addr_start)[0..(i - addr_start)];
+        const addr = try std.fmt.parseInt(u32, addr_str, 16);
+        // Skip to the function name in the second column
+        while (std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        const name_start = i;
+        while (!std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        const name = @intToPtr([*]u8, name_start)[0..(i - name_start)];
+        try syms.add(name, addr);
+        // Skip to the next line
+        while (std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+    }
+    symbol_map = syms;
+    log.logInfo("Done\n");
+}
+
+fn strlen(chars: [*]u8) usize {
+    var i: usize = 0;
+    while (chars[i] != 0) : (i += 1) {}
+    return i;
+}
+
+fn logTraceAddress(addr: usize) void {
+    var str: []const u8 = "(no symbols available)"[0..];
+    if (symbol_map) |syms| {
+        str = syms.search(addr) orelse "?????"[0..];
+    }
+    log.logError("{x}: {}\n", addr, str);
+}
 
 pub fn panic(trace: ?*builtin.StackTrace, comptime format: []const u8, args: ...) noreturn {
     @setCold(true);
-    arch.disableInterrupts();
-    log.logInfo("KERNEL PANIC\n");
-    log.logInfo(format, args);
-    log.logInfo("HALTING\n");
+    log.logError("Kernel panic: " ++ format ++ "\n", args);
+    if (trace) |trc| {
+        var last_addr: u64 = 0;
+        for (trc.instruction_addresses) |ret_addr| {
+            if (ret_addr != last_addr) logTraceAddress(ret_addr);
+            last_addr = ret_addr;
+        }
+    } else {
+        const first_ret_addr = @returnAddress();
+        var last_addr: u64 = 0;
+        var it = std.debug.StackIterator.init(first_ret_addr);
+        while (it.next()) |ret_addr| {
+            if (ret_addr != last_addr) logTraceAddress(ret_addr);
```

Comment:
Surly you don't need this comparison as the iterator will return null for the last case?

Reply:
I added this check so that it wouldn't print loads of duplicate function names in the case of a recursive function, I wouldn't be against removing it though.